developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,255 @@
+buffer.
+
+Similarly to the data, streaming buffer position updates are currently also
+transferred via a dedicated VirtQueue. Both the host and the guest are
+configured to use IPC messages for position updates. Buffers for VirtIO
+VirtQueues are always provided by guests, therefore there should always be a
+position update buffer waiting on the host side to send a position update
+message as soon as one arrives from the DSP.
+
+Upon reception of the STREAM_TRIG_STOP IPC message the vhost driver updates
+the routing information and deactivates the virtual PCM pipeline on the host.
+This operation requires particular care, because in the present state, just
+calling
+
+	soc_dpcm_runtime_update();
+	snd_soc_runtime_deactivate();
+
+doesn't deactivate the pipeline and therefore doesn't allow runtime PM to
+suspend the interface, which then breaks following activation attempts. The
+current VirtIO implementation contains a fix for that, but that fix might turn
+out to be incorrect. The difficulty in establishing a correct fix is, that so
+far it hasn't been possible to find a test case for
+snd_soc_runtime_deactivate().
+
+3.3. Deactivation
+-----------------
+
+When the guest isn't actively using audio, we want to allow the host to
+runtime suspend the DSP. That usually means switching the DSP off. Next time the
+guest wants to activate audio it will have to perform initialisation again,
+including sending its topology to the host. This happens automatically as long
+as the guest runtime suspends too or performs the initialisation cycle in some
+other way, by unloading and reloading drivers, or by rebooting.
+
+With automatic runtime suspending on the host and the guest, it can happen,
+that one of them got suspended, whereas the other one is still running. A
+problematic case would be, if the host got suspended, while the guest was still
+active, because then, when the host resumes, it will also bring up the DSP and
+only send the host topology part to it. The guest however wasn't suspended,
+therefore it won't send its topology to the host. To avoid this the automatic
+runtime suspend timeout on the guest is set to a half of the host's timeout.
+
+The opposite case also requires special handling: when the host is active and
+the guest got suspended and is sending its topology to the host. If the host
+then forwards that topology to the DSP as usual, the DSP firmware will detect
+component duplication and enter an erroneous state. These cases also cannot be
+reliably avoided, because it's always possible, that while the guest's audio is
+idle, the host's audio is still active, e.g. being in use locally. To handle
+such cases correctly, we implement a reset count. The host counts DSP resets,
+and when the guest is resuming, it requests a reset status, using a newly added
+SOF_IPC_PM_VFE_RESET_STATUS IPC message. If the guest identifies, that the host
+hasn't been reset since the guest's last suspend, it skips sending the parsed
+topology to it.
+
```

Comment:
My expectation is that the guest would also send a PM IPC to the DSP to indicate its about to enter D3, the host could use this to track state ?

Reply:
@lgirdwood yes, I guess, the host could use that too. But the point is, that the guest has to know, whether the DSP was suspended since it last was using it. Otherwise it will send the topology to the host when it resumes, even if the DSP wasn't suspended, and therefore still has the complete topology.

Reply:
host owns all DSP PM, and will only suspend if all guests are also suspended (guests send the PM IPC - host should also delete guest pipelines at this point). This means that on guest resume code flow is exactly the same as host.